### ISSUE TITLE
`ForeignKey` method documentation incorrectly states that Column strings use table name, but use class name at runtime

### ISSUE THREAD

**Issue by tomthorogood** [sqlalchemy/sqlalchemy#5042]
The [documentation for ForeignKey](https://docs.sqlalchemy.org/en/13/core/constraints.html?highlight=foreignkey#sqlalchemy.schema.ForeignKey) reads:

> column – A single target column for the key relationship. A Column object **or a column name as a string: tablename.columnkey or schema.tablename.columnkey.** columnkey is the key which has been assigned to the column (defaults to the column name itself), unless link_to_name is True in which case the rendered name of the column is used.

TL;DR: in reality it is sometimes attempting to resolve by the class name, *not* the table name. We are in the middle of a refactor, and after renaming our model classes, `ForeignKey` calls no longer work, with the traceback:

```
[...]
../env/lib/python3.7/site-packages/sqlalchemy/orm/interfaces.py:196: in init
    self.do_init()
../env/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1878: in do_init
    self._process_dependent_arguments()
../env/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1907: in _process_dependent_arguments
    setattr(self, attr, attr_value())
../env/lib/python3.7/site-packages/sqlalchemy/ext/declarative/clsregistry.py:294: in __call__
    x = eval(self.arg, globals(), self._dict)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   AttributeError: 'Table' object has no attribute 'other_model_id'
```

Before our refactor, we had the following setup:

```
Base = declarative_base()

class OtherModel(Base):
    __tablename__ = "OtherModel"
    customer_id = Column('other_model_id', Integer, primary_key=True)

class MyModel(Base):
    __tablename__ = "MyModel"
    customer_id = Column('customer_id', Integer, ForeignKey('OtherModel.other_model_id'))
```

The above example works just fine.

However, on a refactor, we wanted to softly stop depending on the declarative base subclasses for business logic, and use a `pydantic` model as an application layer go-between (to allow us more flexibility and add JSON Schema support), so we have something like this instead:

```
Base = declarative_base()

class OtherModel(pydantic.BaseModel):
   other_model_id: int

class OtherModelStorage(Base):
    __tablename__ = "OtherModel"
    other_model_id = Column('other_model_id', Integer, primary_key=True)

class MyModel(pydantic.BaseModel):
    other_model_id: int

class MyModelStorage(Base):
    __tablename__ = "MyModel"
    other_model_id = Column('customer_id', Integer, ForeignKey('OtherModel.other_model_id'))
```

we get the error posted above. As an experiment, I also changed the `ForeignKey` call to look like: `ForeignKey('OtherModelStorage.other_model_id'))` 

This resulted in a different error:

```
sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'MyModel.other_model_id' could not find table 'OtherModelStorage' with which to generate a foreign key to target column 'other_model_id'
```

This means that *sometimes* the column string is being interpreted as a column on a table, and other times it's being used to reference an actual class. There doesn't seem to be any way to know which, or any way to resolve this without either renaming everything back the way it was, or renaming our tables so that they don't share a name with an actual object name. Neither option is really great, and it doesn't seem like this should actually be an issue, according to the docs.

I think that if the behavior can't be fixed to be consistent, then the docs should be updated to point out this behavior, so that others can be aware. (The impact being that had I known ahead of time, I could have opted for a different refactor approach; as it is, I have to spend another day undoing my refactor and applying it differently.)

Thank you.

**Comment by zzzeek** [comment]
hi there -

> The [documentation for ForeignKey](https://docs.sqlalchemy.org/en/13/core/constraints.html?highlight=foreignkey#sqlalchemy.schema.ForeignKey) reads:
> 
> > column – A single target column for the key relationship. A Column object **or a column name as a string: tablename.columnkey or schema.tablename.columnkey.** columnkey is the key which has been assigned to the column (defaults to the column name itself), unless link_to_name is True in which case the rendered name of the column is used.
> 
> TL;DR: in reality it is sometimes attempting to resolve by the class name, _not_ the table name. 

That's not correct; here's an example:

```
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.schema import CreateTable

Base = declarative_base()


class AClass(Base):
    __tablename__ = "a_table"

    id = Column(Integer, primary_key=True)
    data = Column(String)


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(Integer, ForeignKey("AClass.id"))

print(CreateTable(B.__table__))

```
output:

`   sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'b.a_id' could not find table 'AClass' with which to generate a foreign key to target column 'id'
`

The error above illustrates that ForeignKey knows nothing about a class named "AClass".  It would only be able to locate the table named "a_table", which it would do by looking inside the `MetaData` collection that both "a_table" and "b" are members of.

The mechanics by which declarative plugs in the "class registry" lookup is limited to the `relationship()` ORM function, and nothing else.

> We are in the middle of a refactor, and after renaming our model classes, `ForeignKey` calls no longer work, with the traceback:
> 
> ```
> [...]
> ../env/lib/python3.7/site-packages/sqlalchemy/orm/interfaces.py:196: in init
>     self.do_init()
> ../env/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1878: in do_init
>     self._process_dependent_arguments()
> ../env/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1907: in _process_dependent_arguments
>     setattr(self, attr, attr_value())
> ../env/lib/python3.7/site-packages/sqlalchemy/ext/declarative/clsregistry.py:294: in __call__
>     x = eval(self.arg, globals(), self._dict)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> >   ???
> E   AttributeError: 'Table' object has no attribute 'other_model_id'

That traceback is coming from relationship(), which is the function that uses the declarative class registry, not ForeignKey(), which does not.   

In the case of relationship() when used with declarative, the string resolution **does** necessarily have to consult both the names of classes as well as the names of tables in the general sense, though it could perhaps be improved to not do so in the case of the referenced class name, but otherwise it is also available for use in the expressions which are passed to arguments like `primaryjoin`, which do often need to refer to Table objects and class objects interchangeably.    The documentation for this string resolution is at https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/relationships.html however it does not point out at the moment the order of resolution between classes and tables (classes take precedence).

> 
> ```
> Base = declarative_base()
> 
> class OtherModel(Base):
>     __tablename__ = "OtherModel"
>     customer_id = Column('other_model_id', Integer, primary_key=True)
> 
> class MyModel(Base):
>     __tablename__ = "MyModel"
>     customer_id = Column('customer_id', Integer, ForeignKey('OtherModel.other_model_id'))
> ```
> 
> The above example works just fine.

Note that the ForeignKey above is matching the table name, not the class name, which happen to be identical in this example.

> 
> However, on a refactor, we wanted to softly stop depending on the declarative base subclasses for business logic, and use a `pydantic` model as an application layer go-between (to allow us more flexibility and add JSON Schema support), so we have something like this instead:
> 
> ```
> Base = declarative_base()
> 
> class OtherModel(pydantic.BaseModel):
>    other_model_id: int
> 
> class OtherModelStorage(Base):
>     __tablename__ = "OtherModel"
>     other_model_id = Column('other_model_id', Integer, primary_key=True)
> 
> class MyModel(pydantic.BaseModel):
>     other_model_id: int
> 
> class MyModelStorage(Base):
>     __tablename__ = "MyModel"
>     other_model_id = Column('customer_id', Integer, ForeignKey('OtherModel.other_model_id'))
> ```
> 

> we get the error posted above.

the above example is not illustrating the use of relationship(), which is where the stack trace is originating from.  the ForeignKey above should be working fine as it refers to the "OtherModel" table name.


 > As an experiment, I also changed the `ForeignKey` call to look like: > `ForeignKey('OtherModelStorage.other_model_id'))`
> 
> This resulted in a different error:
> 
> ```
> sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'MyModel.other_model_id' could not find table 'OtherModelStorage' with which to generate a foreign key to target column 'other_model_id'

Correct, that is the error that ForeignKey raises when it can't locate a particular table name, which you can see is the same error emitted by my demonstration script.  it is not the same exception class,  message, or stack trace you are getting from the error message you are having a problem with, as that's a totally different error that emits from the ORM relationship function.


> ```
> 
> This means that _sometimes_ the column string is being interpreted as a column on a table, and other times it's being used to reference an actual class. There doesn't seem to be any way to know which, or any way to resolve this without either renaming everything back the way it was, or renaming our tables so that they don't share a name with an actual object name. Neither option is really great, and it doesn't seem like this should actually be an issue, according to the docs.

So to sum up it seems like you are having problems with calls to relationship(), not ForeignKey(), which are two totally different features; one is from the ORM and links ORM-mapped classes to each other, the other is from Core and links Table objects to each other.     There's no example code showing what you're doing with relationship() but generally it looks like you have a relationship() that refers to a name that links to a Table, when you intend for it to link to a class.  The string resolution in relationship() necessarily resolves to both classes and tables as it is used to assist in expressions like "primaryjoin" which often do refer to Table objects directly.  So if you are changing the names of classes, you need to change the names you are using in relationship() to refer to them.

**Comment by tomthorogood** [comment]
Thank you so much for the insightful response. (I can't believe I didn't catch that it was `relationship`, not `ForeignKey` throwing the error.) 

> generally it looks like you have a relationship() that refers to a name that links to a Table, when you intend for it to link to a class

With the above example, including the call to `relationship`, it looks like:

```
class MyModelStorage(Base):
    __tablename__ = "MyModel"
    other_model_id = Column('customer_id', Integer, ForeignKey('OtherModel.other_model_id'))
    other_model = relationship('OtherModelStorage')
```

I believe this is what you recommended above. However, it's a large code base, and I cannot say with certainty that I didn't perhaps miss one during the refactor.

I will go through and make certain that calls to `ForeignKey` use the *table* name, and calls to `relationship` use the *class* name. Thank you for helping me untangle the differences there.

🤞🏻

**Comment by tomthorogood** [comment]
It looked like, as you recommended, the issue came from referring to the table in a couple `primaryjoin` parameters. 

Thanks again!

**Comment by zzzeek** [comment]
great, then we're good
